Signal/src/ViewControllers/MessagesViewController.m

```diff
@@ -2020,48 +2031,83 @@ - (void)tappedInvalidIdentityKeyErrorMessage:(TSInvalidIdentityKeyErrorMessage *
+                    [errorMessage acceptNewIdentityKey];
+                    if ([errorMessage isKindOfClass:[TSInvalidIdentityKeySendingErrorMessage class]]) {
+                        [self.messageSender
+                            resendMessageFromKeyError:(TSInvalidIdentityKeySendingErrorMessage *)errorMessage
+                            success:^{
+                                DDLogDebug(@"%@ Successfully resent key-error message.", self.tag);
+                            }
+                            failure:^(NSError *_Nonnull error) {
+                                DDLogError(@"%@ Failed to resend key-error message with error:%@", self.tag, error);
+                            }];
+                    }
+                }];
     [actionSheetController addAction:acceptSafetyNumberAction];
     
     [self presentViewController:actionSheetController animated:YES completion:nil];
 }
 
+- (void)tappedUnknownContactBlockOfferMessage:(OWSUnknownContactBlockOfferMessage *)errorMessage
+{
+    NSString *displayName = [self.contactsManager displayNameForPhoneIdentifier:errorMessage.contactId];
+    NSString *title =
+        [NSString stringWithFormat:NSLocalizedString(@"BLOCK_OFFER_ACTIONSHEET_TITLE_FORMAT",
+                                       @"Title format for action sheet that offers to block an unknown user."
+                                       @"Embeds {{the unknown user's name or phone number}}."),
+                  [BlockListUIUtils formatDisplayNameForAlertTitle:displayName]];
+
+    UIAlertController *actionSheetController =
+        [UIAlertController alertControllerWithTitle:title message:nil preferredStyle:UIAlertControllerStyleActionSheet];
+
+    UIAlertAction *dismissAction = [UIAlertAction actionWithTitle:NSLocalizedString(@"TXT_CANCEL_TITLE", @"")
+                                                            style:UIAlertActionStyleCancel
+                                                          handler:nil];
+    [actionSheetController addAction:dismissAction];
+
+    UIAlertAction *blockAction =
+        [UIAlertAction actionWithTitle:NSLocalizedString(@"BLOCK_OFFER_ACTIONSHEET_BLOCK_ACTION",
+                                           @"Action sheet that will block an unknown user.")
+                                 style:UIAlertActionStyleDestructive
+                               handler:^(UIAlertAction *_Nonnull action) {
+                                   DDLogInfo(@"%@ Blocking an unknown user.", self.tag);
+                                   [self.blockingManager addBlockedPhoneNumber:errorMessage.contactId];
+                                   // Delete the block offer.
+                                   [self.storageManager.dbConnection
+                                       readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
+                                           [errorMessage removeWithTransaction:transaction];
```

Comment:
I haven't tried to use this yet, but it seems like the block offer would reappear when the unknown-user sends a new message, which seems undesirable, since the receiving user has stated the don't want to block the sending user.